### PR TITLE
Add float8 FakeQuantizeConfig and FakeQuantizer

### PR DESCRIPTION
**Pull request by andrewor14**
**Summary:** This commit adds a QAT path for float8, using the same primitives as `torchao.quantization.Float8Tensor` targeting the following PTQ configs:
- `Float8DynamicActivationFloat8WeightConfig`
- `Float8ActivationInt4WeightConfig`

Usage:
```
from torchao.quantization.granularity import PerRow
from torchao.quantization.qat import quantize_, QATConfig

base_config = Float8DynamicActivationFloat8WeightConfig(
    torch.float8_e4m3fn, PerRow(),
)
quantize_(model, QATConfig(base_config, step="prepare"))
quantize_(model, QATConfig(base_config, step="convert"))
```

OR

```
from torchao.quantization.granularity import PerRow
from torchao.quantization.qat import (
    Float8FakeQuantizeConfig,
    QATConfig,
    quantize_,
)

dtype = torch.float8_e4m3fn
granularity = PerRow()
quantize_(model, QATConfig(
    activation_config=Float8FakeQuantizeConfig(dtype, granularity),
    weight_config=Float8FakeQuantizeConfig(dtype, granularity),
    step="prepare",
)

# convert (same as above, not shown)
```

**Test Plan:**
```
python test/quantization/test_qat.py -k test_float8_fake_quantize_config
python test/quantization/test_qat.py -k test_float8_fake_quantize
python test/quantization/test_qat.py -k test_quantize_api_fp8_fp8
python test/quantization/test_qat.py -k test_quantize_api_fp8_int4
```

Identical outputs between normal bf16 and QAT fine-tuning for both fp8-fp8 and fp8-int4, reproduced on Llama3.1 using [this unsloth notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-Alpaca.ipynb). Loss curves also overlap almost exactly (not shown):
```
<|begin_of_text|>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Continue the fibonnaci sequence.

### Input:
1, 1, 2, 3, 5, 8

### Response:
13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817, 39088169, 632459
```

### CODE REVIEW

File: torchao/quantization/quant_primitives.py [hunk 1]
```diff
@@ -2295,10 +2297,12 @@ def _quantize_affine_float8(
...
     max_value = torch.finfo(float8_dtype).max
     tensor_clamped = tensor_scaled.clamp(min=-max_value, max=max_value)
-    fp8_tensor = tensor_clamped.to(float8_dtype)
-    return fp8_tensor
+    if cast_to_float8_dtype:
+        tensor_clamped = tensor_clamped.to(float8_dtype)
+    return tensor_clamped
 
 
+# TODO: don't register as custom op?
```
Comment (andrewor14) [inline torchao/quantization/quant_primitives.py:2305]: @jerryzh168 I'm seeing this warning. Maybe should also skip registering this custom op?
```
/home/andrewor/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/torch/autograd/graph.py:824: UserWarning: torchao::dequantize_affine_float8: an autograd kernel was not registered to the Autograd key(s) but we are trying to backprop through it. This may lead to silently incorrect behavior. This behavior is deprecated and will be removed in a future version of PyTorch. If your operator is differentiable, please ensure you have registered an autograd kernel to the correct Autograd key (e.g. DispatchKey::Autograd, DispatchKey::CompositeImplicitAutograd). If your operator is not differentiable, or to squash this warning and use the previous behavior, please register torch::CppFunction::makeFallthrough() to DispatchKey::Autograd. (Triggered internally at /pytorch/torch/csrc/autograd/autograd_not_implemented_fallback.cpp:62.)
```

File: torchao/quantization/qat/fake_quantize_config.py [hunk 2]
```diff
@@ -302,18 +346,52 @@ def _infer_fake_quantize_configs(
...
+            dtype=base_config.activation_dtype,
+            granularity=act_granularity,
+            hp_value_lb=base_config.activation_value_lb,
+            hp_value_ub=base_config.activation_value_ub,
+        )
+        weight_config = Float8FakeQuantizeConfig(
+            dtype=base_config.weight_dtype,
+            granularity=weight_granularity,
+        )
+    elif isinstance(base_config, Float8ActivationInt4WeightConfig):
```
Comment (andrewor14) [inline torchao/quantization/qat/fake_quantize_config.py]: @jerryzh168 can you confirm these config settings?

File: torchao/quantization/quant_primitives.py [hunk 3]
```diff
@@ -2295,10 +2297,12 @@ def _quantize_affine_float8(
...
 @_register_custom_op(quant_lib, False)
 def _dequantize_affine_float8(
```
Comment (andrewor14) [inline torchao/quantization/quant_primitives.py:2298]: @jerryzh168 I'm seeing this warning. Maybe should also skip registering this custom op?
```
/home/andrewor/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/torch/autograd/graph.py:824:
UserWarning: torchao::dequantize_affine_float8: an autograd kernel was not registered to the
Autograd key(s) but we are trying to backprop through it. This may lead to silently incorrect behavior.
This behavior is deprecated and will be removed in a future version of PyTorch.
If your operator is differentiable, please ensure you have registered an autograd kernel to
the correct Autograd key (e.g. DispatchKey::Autograd, DispatchKey::CompositeImplicitAutograd).
If your operator is not differentiable, or to squash this warning and use the previous behavior,
please register torch::CppFunction::makeFallthrough() to DispatchKey::Autograd.
(Triggered internally at /pytorch/torch/csrc/autograd/autograd_not_implemented_fallback.cpp:62.)
```

File: torchao/quantization/qat/fake_quantize_config.py [hunk 4]
```diff
@@ -302,18 +346,52 @@ def _infer_fake_quantize_configs(
...
+        if base_config.VERSION == 2:
+            # targeting fbgemm kernel
+            weight_config = IntxFakeQuantizeConfig(
+                dtype=torch.int4,
+                group_size=base_config.group_size,
+                is_symmetric=True,
+            )
+        else:
+            # targeting tinygemm kernel
+            assert base_config.VERSION == 1
```
Comment (vkuzo) [inline torchao/quantization/qat/fake_quantize_config.py]: just support version 2, to minimize complexity?

**Comment by vkuzo** [comment]
can the test plan include training a real model and verifying loss converges

**Comment by andrewor14** [comment]
> can the test plan include training a real model and verifying loss converges

Yes this is in progress

File: test/quantization/test_qat.py [hunk 5]
```diff
@@ -58,10 +59,11 @@
 from torchao.quantization.qat.utils import (
     _fake_quantize_per_channel_group,
     _fake_quantize_per_token,
-    _Float8RowwiseFakeQuantize,
     _get_qmin_qmax,
 )
 from torchao.quantization.quant_api import (
+    Float8ActivationInt4WeightConfig,
```
Comment (jerryzh168) [inline test/quantization/test_qat.py]: nit: we just renamed this one

File: test/quantization/test_qat.py [hunk 6]
```diff
@@ -1912,6 +1898,78 @@ def test_qat_api_deprecation(self):
...
+        fq_config = Float8FakeQuantizeConfig(dtype, granularity)
+        fake_quantizer = Float8FakeQuantizer(fq_config)
+        torch.manual_seed(self.SEED)
+        x = torch.randn(32, 64)
+        out = fake_quantizer(x)
+        out_expected = Float8Tensor.to_float8(x, dtype, granularity).dequantize()
+        sqnr = compute_error(out, out_expected)
+        self.assertGreater(sqnr, 16)
+
+    @parameterized.expand([(PerRow(),), (PerTensor(),)])
```
Comment (jerryzh168) [inline test/quantization/test_qat.py]: nit: why not use https://github.com/pytorch/ao/blob/1dca63818d7e7cd96b403657aa0dde6170a560da/test/quantization/quantize_/workflows/float8/test_float8_tensor.py#L66

File: test/quantization/test_qat.py [hunk 7]
```diff
@@ -1912,6 +1898,78 @@ def test_qat_api_deprecation(self):
...
+        m = M().to(torch.bfloat16).cuda()
+        example_inputs = (m.example_inputs()[0].to(torch.bfloat16).cuda(),)
+        base_config = Float8ActivationInt4WeightConfig(group_size=128)
+        try:
+            quantize_(m, QATConfig(base_config, step="prepare"))
+            quantize_(m, QATConfig(base_config, step="convert"))
+            m(*example_inputs)
+        except ImportError as e:
+            if "fbgemm-gpu-genai" in str(e):
+                self.skipTest("fbgemm-gpu-genai not available")
```
Comment (jerryzh168) [inline test/quantization/test_qat.py]: nit: we can skip the test when fbgemm-gpu-genai is not installed: https://github.com/pytorch/ao/blob/1dca63818d7e7cd96b403657aa0dde6170a560da/test/quantization/quantize_/workflows/int4/test_int4_preshuffled_tensor.py#L45

File: test/quantization/test_qat.py [hunk 8]
```diff
@@ -1912,6 +1898,78 @@ def test_qat_api_deprecation(self):
...
+            quantize_(model, QATConfig(Float8ActivationInt4WeightConfig(), step="convert"))
+        """
+        torch.manual_seed(self.SEED)
+        m = M().to(torch.bfloat16).cuda()
+        example_inputs = (m.example_inputs()[0].to(torch.bfloat16).cuda(),)
+        base_config = Float8ActivationInt4WeightConfig(group_size=128)
+        try:
+            quantize_(m, QATConfig(base_config, step="prepare"))
+            quantize_(m, QATConfig(base_config, step="convert"))
+            m(*example_inputs)
```
Comment (jerryzh168) [inline test/quantization/test_qat.py]: should this happen between prepare and convert as well

File: torchao/quantization/qat/fake_quantize_config.py [hunk 9]
```diff
@@ -302,18 +346,45 @@ def _infer_fake_quantize_configs(
...
+            granularity=act_granularity,
+            hp_value_lb=base_config.activation_value_lb,
+            hp_value_ub=base_config.activation_value_ub,
+        )
+        weight_config = Float8FakeQuantizeConfig(
+            dtype=base_config.weight_dtype,
+            granularity=weight_granularity,
+        )
+    elif isinstance(base_config, Float8DynamicActivationInt4WeightConfig):
+        act_config = Float8FakeQuantizeConfig(
```
Comment (jerryzh168) [inline torchao/quantization/qat/fake_quantize_config.py:379]: one thing we should pay extra attention here is whether the simulation works for int4 preshuffled tensor as well I think, we need some numerics testing to make sure

File: torchao/quantization/quant_primitives.py [hunk 10]
```diff
@@ -2200,13 +2200,14 @@ def _choose_scale_float8(
...
     quant_max = torch.finfo(float8_dtype).max
-    # only tensorwise scaling is supported for now:
     if len(block_size) == 0:
+        # tensorwise
         max_abs = tensor.abs().max()
         if hp_value_lb is not None or hp_value_ub is not None:
             max_abs = torch.clamp(max_abs, min=hp_value_lb, max=hp_value_ub)
         scale = max_abs / quant_max
     else:
+        # rowwise
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py]: this is not necessarily rowwise I think, I believe this includes all granularities, and the len(block_size) == 0 is more of special case for tensorwise quant, I'm not sure where it comes from and whether it's needed, we could try to trace it and see if it can be removed as well to reduce complexity

File: torchao/quantization/quant_primitives.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:2298]: oh, we didn't know this would be a problem, we can do https://github.com/pytorch/ao/blob/1dca63818d7e7cd96b403657aa0dde6170a560da/torchao/quantization/quant_primitives.py#L361 for now I think, this is added from intel for https://github.com/pytorch/ao/pull/2565 I believe and it's broken now anyways

Comment (andrewor14) [reply torchao/quantization/quant_primitives.py:2298]: ok, will relax in a separate PR

File: test/quantization/test_qat.py [hunk 11]
```diff
@@ -1833,6 +1826,82 @@ def test_qat_api_convert_no_quantization(self):
...
+            ValueError, "Please specify the granularity object instead of the class"
+        ):
+            Float8FakeQuantizeConfig(granularity=PerRow)
+        with self.assertRaisesRegex(
+            ValueError, "Expected PerRow or PerTensor granularity"
+        ):
+            Float8FakeQuantizeConfig(granularity=PerToken())
+
+    @parametrize("granularity", [PerTensor(), PerRow()])
+    def test_float8_fake_quantize(self, granularity: Granularity):
```
Comment (jerryzh168) [inline test/quantization/test_qat.py:1851]: can you add a same test for fp8_int4?

Comment (andrewor14) [reply test/quantization/test_qat.py:1851]: added some sqnr comparison against PTQ fp8_int4

File: torchao/quantization/qat/fake_quantize_config.py (hunk 9 above)
Comment (andrewor14) [reply torchao/quantization/qat/fake_quantize_config.py:379]: added an fp8-int4 numerics test
